File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
import math

def solution(maps):    
    return bfs(maps)


def bfs(maps):
    n = len(maps)
    m = len(maps[0])
    queue = [[0, 0]]    
    visited = set([(0, 0)])    
    distances = {(0, 0):1}
    
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    
    while queue:
        cr, cc = queue.pop(0)
        
        if cr == n - 1 and cc == m - 1:
            return distances[(cr, cc)]
        
        for d in range(4):
            nr, nc = cr + dr[d], cc + dc[d]
            if 0 <= nr < n and 0 <= nc < m and maps[nr][nc] == 1 and (nr, nc) not in visited:            
                visited.add((nr, nc))
                distances[(nr, nc)] = distances[(cr, cc)] + 1
                queue.append([nr, nc])            
    
    return -1
    
                

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int[] dr = {-1, 1, 0, 0};
        int[] dc = {0, 0, -1, 1};
        int N = maps.length;
        int M = maps[0].length;        
        boolean[][] visited = new boolean[N][M];
        
        Deque<int[]> queue = new ArrayDeque<>();
        queue.add(new int[] {0, 0, 1}); // r, c, 지난 칸수
        visited[0][0] = true;
        
        while(!queue.isEmpty()){
            int cr = queue.peek()[0];
            int cc = queue.peek()[1];
            int cBlock = queue.peek()[2];
            queue.poll();
            
            if(cr == N - 1 && cc == M - 1){
                return cBlock;
            }
            
            for(int d = 0; d < 4; d++){
                int nr = cr + dr[d];
                int nc = cc + dc[d];
                
                if(nr < 0 || N <= nr || nc < 0 || M <= nc || visited[nr][nc] || maps[nr][nc] == 0) continue;
                queue.add(new int[] {nr, nc, cBlock + 1});
                visited[nr][nc] = true;
            }
        }
        
        return -1;
    }
}